### PR TITLE
Faster tree builder

### PR DESCRIPTION
**Pull request by mqzry**
Reintroduces one of the optimisations mentioned in https://github.com/google/xi-editor/issues/583.
This optimisation uses a height sorted stack to push a string on an existing tree.

I also added some benches. Benches with 'stack' in the name are with the new method:

`test build_tree_few_big_lines           ... bench:      90,491 ns/iter (+/- 6,478)`
`test build_tree_few_big_lines_stack ... bench:      80,863 ns/iter (+/- 4,891)`

`test build_tree_short_lines                     ... bench:     322,070 ns/iter (+/- 21,143)`
`test build_tree_short_lines_stack        	... bench:     209,952 ns/iter (+/- 9,921)`

`test build_tree_triangle                           ... bench:     503,447 ns/iter (+/- 36,269)`
`test build_tree_triangle_stack             	... bench:     299,227 ns/iter (+/- 23,092)`

### CODE REVIEW

**Comment by cmyr** [comment]
Cool, those are significant gains. Probably won't get around to giving this a thorough review for a few days, but happy to have you plugging away at perf stuff!

**Comment by mqzry** [comment]
I feel like there should be some tests.

I will add on sunday a test to see if the new method yields exectly the same rope as the old method.

Let me know if there are any other scenarios I should.

File: rust/rope/src/tree.rs [hunk 1]
```diff
@@ -478,7 +471,29 @@ impl<N: NodeInfo> TreeBuilder<N> {
     pub fn push(&mut self, n: Node<N>) {
         match self.0.take() {
             None => self.0 = Some(n),
-            Some(buf) => self.0 = Some(Node::concat(buf, n))
+            Some(buf) => self.0 = Some(Node::concat(buf, n)),
+        }
+    }
+
+    pub fn push_leaves(&mut self, leaves: Vec<N::L>) {
```
Comment (raphlinus) [inline rust/rope/src/tree.rs]: Being a pub fn, this should have at least a short docstring. I know that's not upheld in the existing code, but trying to raise the standard...

File: rust/rope/src/tree.rs [hunk 2]
```diff
@@ -478,7 +471,29 @@ impl<N: NodeInfo> TreeBuilder<N> {
...
+        let mut stack: Vec<Vec<Node<N>>> = Vec::new();
+        for leaf in leaves {
```
Comment (raphlinus) [inline rust/rope/src/tree.rs]: I found this a bit hard to follow (and I probably wrote the original logic). What I think would be most helpful is a description of invariants. I _believe_ the following is true, and could be adapted into comments:

The stack is a list of list of nodes. Each list of nodes is of uniform height, and of length at least one and at most MAX_CHILDREN - 1. The heights are strictly decreasing.

Having gone through it, this looks right.

File: rust/rope/benches/rope.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
+// Copyright 2017 Google Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (jansol) [reply rust/rope/benches/rope.rs]: Since this is being changed for the rest of the project already in #778 I think this should also be updated to the new line.

[the remark this answers is not in the document]
Comment (mqzry) [reply rust/rope/benches/rope.rs]: Will change it to the new line as soon as I get to it

File: rust/rope/src/tree.rs (hunk 2 above)
Comment (mqzry) [reply rust/rope/src/tree.rs]: Those invariants are correct. I added a description of those.

I was/am doubting between writing just a short doc string documenting the invariants and a full blown explanation why these invariants lead to amortized linear time.

**Comment by mqzry** [comment]
Ready for next review or merge

File: rust/rope/src/tree.rs (hunk 2 above)
Comment (jansol) [reply rust/rope/src/tree.rs]: More detailed documentation almost never hurts. If you're not sure I'd say write both, much like you would write a long git commit message, i.e. a short oneliner, an empty line and a long description.

Comment (jansol) [reply rust/rope/src/tree.rs]: Aaaand I see that's exactly what you did. These github change views can be confusing at times.

File: rust/rope/src/rope.rs [hunk 4]
```diff
@@ -677,6 +680,26 @@ impl TreeBuilder<RopeInfo> {
             s = &s[splitpoint..];
         }
     }
+
+    pub fn push_str_stacked(&mut self, s: &str) {
```
[the remark this answers is not in the document]
Comment (mqzry) [reply rust/rope/src/rope.rs]: This is only used for testing and benching. As far as I can tell, #[cfg(test)] also works for `cargo bench` so I've added it.

File: rust/rope/src/rope.rs [hunk 5]
```diff
@@ -822,7 +845,13 @@ impl<'a> Iterator for LinesRaw<'a> {
             if self.fragment.is_empty() {
                 match self.inner.next() {
                     Some(chunk) => self.fragment = chunk,
-                    None => return if result.is_empty() { None } else { Some(result) }
+                    None => {
```
[the remark this answers is not in the document]
Comment (mqzry) [reply rust/rope/src/rope.rs]: Sorry for that. Will rework the commit next time, if that is the case. Would love to see rustfmt incorporated indeed.

**Comment by mqzry** [comment]
@raphlinus, this needs to be approved again. Incorporated the small nits of cmyr.

**Comment by cmyr** [comment]
Awesome, thanks for the comments and sorry for the delay. For what it's worth I've been spending a lot more time in the rope crate recently, I should be better able to review changes going forward.

**Comment by cmyr** [comment]
@mqzry sounds good!
